File: lagomtest/household-api/src/main/java/household/api/Household.java
package household.api;

import java.util.List;

import javax.annotation.concurrent.Immutable;

import com.fasterxml.jackson.annotation.JsonCreator;
import com.fasterxml.jackson.databind.annotation.JsonDeserialize;

@Immutable
@JsonDeserialize
public class Household {

	private final String id;
	private final String name;
	private final List<InventoryItem> inventoryItems;
	
	@JsonCreator
	public Household(String id, String name, List<InventoryItem> inventoryItems) {
		this.id = id;
		this.name = name;
		this.inventoryItems = inventoryItems;
	}

	/**
	 * @return this instance's unique id
	 */
	public String id() {
		return id;
	}

	/**
	 * @return the {@code Household}'s display name
	 */
	public String name() {
		return name;
	}

	/**
	 * @return this {@code Household}'s list of {@code InventoryItem}s
	 */
	public List<InventoryItem> inventoryItems() {
		return inventoryItems;
	}
}

File: lagomscala/household-api/src/main/scala/lagomscala/household/api/Household.scala
package lagomscala.household.api

import java.time.Instant
import java.util.UUID

import play.api.libs.json.{Format, Json}
import valueobjects.{Amount, EAN}

import scala.collection.immutable.Seq

case class InventoryItem(id: Option[UUID],
                         productId: EAN,
                         originalAmount: Amount,
                         remainingFraction: Float,
                         obtained: Option[Instant],
                         bestBefore: Option[Instant]) {
}

object InventoryItem {

  implicit val format: Format[InventoryItem] = Json.format[InventoryItem]

}

/**
  * The Household Aggregate Root
  *
  * @param id
  * @param name
  * @param owner
  * @param inventoryItems
  */
case class Household(
                      id: Option[UUID],
                      name: Option[String],
                      owner: String,
                      inventoryItems: Seq[InventoryItem]) {

  def this(owner: String) = this(None, None, owner, Seq.empty)

}

object Household {

  implicit val format: Format[Household] = Json.format[Household]

}

